### PR TITLE
Add WS support for new query pipeline

### PR DESCRIPTION
**Pull request by vasukalariya**
# Description
Summarize the change.
Use the existing API pipeline and create WS responses as required.

Currently whatever the number of records are returned will be used to display the results similar to how its present in the API response.

TODO:
- Use the search result to create send totalMatched and other stats correctly. It should also account for is search exhausted etc. and based on it the relation should be displayed. This would be done in a later PR.

Fixes #<issue-number> (link all the GitHub issues this addresses)

# Testing
Describe how you tested this code. How can the reviewers reproduce your tests?
Tested on queries like
`city=Boston | sort country`
`city=Boston | where first_name="Alanis"`

image

image

# Checklist:
Before marking your pull request as ready for review, complete the following.

- [ ] I have self-reviewed this PR.
- [ ] I have removed all print-debugging and commented-out code that should not be merged.
- [ ] I have added sufficient comments in my code, particularly in hard-to-understand areas.
- [ ] I have formatted the code, if applicable. For Go, I have run `goimports -w .`.

### CODE REVIEW

File: pkg/segment/query/querystatus.go [hunk 1]
```diff
@@ -113,6 +114,7 @@ type RunningQueryState struct {
 	nodeResult               *structs.NodeResult
 	totalRecsToBeSearched    uint64
 	AllColsInAggs            map[string]struct{}
+	NewPipeLineResponse      *structs.PipeSearchResponseOuter
```
Comment (nkunal) [inline pkg/segment/query/querystatus.go:117]: very soon the `NewPipeLine` is going to be not new anymore. And maybe later we come up with some other new pipeline, so perhaps we should rename this var. maybe `PipeResp` or something

File: pkg/segment/query/querystatus.go [hunk 2]
```diff
@@ -925,3 +937,43 @@ func LogGlobalSearchErrors(qid uint64) {
 		putils.LogUsingLevel(errInfo.LogLevel, "qid=%v, %v, Count: %v, ExtraInfo: %v", qid, errMsg, errInfo.Count, errInfo.Error)
 	}
 }
+
+func SetNewQueryPipelineResponse(response *structs.PipeSearchResponseOuter, qid uint64) error {
```
Comment (nkunal) [inline pkg/segment/query/querystatus.go:941]: same comment as above, rename this func based on what name you give for the var above

File: pkg/segment/query/querystatus.go [hunk 3]
```diff
@@ -925,3 +937,43 @@ func LogGlobalSearchErrors(qid uint64) {
...
+	arqMapLock.RLock()
+	rQuery, ok := allRunningQueries[qid]
+	arqMapLock.RUnlock()
+	if !ok {
+		log.Errorf("SetQueryResponse: qid %+v does not exist!", qid)
```
Comment (nkunal) [inline pkg/segment/query/querystatus.go:946]: fix func name

File: pkg/segment/query/querystatus.go [hunk 4]
```diff
@@ -925,3 +937,43 @@ func LogGlobalSearchErrors(qid uint64) {
...
+		return fmt.Errorf("qid does not exist")
```
Comment (nkunal) [inline pkg/segment/query/querystatus.go:947]: add qid here

File: pkg/segment/query/querystatus.go [hunk 5]
```diff
@@ -925,3 +937,43 @@ func LogGlobalSearchErrors(qid uint64) {
...
+
+	rQuery.StateChan <- &QueryStateChanData{
+		StateName:       QUERY_UPDATE,
+		QueryUpdate:     &QueryUpdate{QUpdate: QUERY_UPDATE_LOCAL},
+		PercentComplete: 100,
+	}
+	return nil
+}
+
+func GetNewQueryPipelineResponse(qid uint64) *structs.PipeSearchResponseOuter {
```
Comment (nkunal) [inline pkg/segment/query/querystatus.go:967]: same comment as above, rename this func based on what name you give for the var above

File: pkg/segment/query/querystatus.go [hunk 6]
```diff
@@ -925,3 +937,43 @@ func LogGlobalSearchErrors(qid uint64) {
...
+	arqMapLock.RLock()
+	rQuery, ok := allRunningQueries[qid]
+	arqMapLock.RUnlock()
+	if !ok {
+		log.Errorf("GetQueryResponse: qid %+v does not exist!", qid)
```
Comment (nkunal) [inline pkg/segment/query/querystatus.go:972]: fix func name

File: pkg/segment/segexecution.go [hunk 7]
```diff
@@ -459,11 +461,54 @@ func ExecuteAsyncQuery(root *structs.ASTNode, aggs *structs.QueryAggregators, qi
...
+		if config.IsNewQueryPipelineEnabled() {
+			_ = executeNewPipelineQueryInternal(root, aggs, qid, qc)
+		} else {
+			_ = executeQueryInternal(root, aggs, qid, qc, rQuery)
+		}
 	}()
 	return rQuery.StateChan, nil
 }
 
+func executeNewPipelineQueryInternal(root *structs.ASTNode, aggs *structs.QueryAggregators, qid uint64, qc *structs.QueryContext) *structs.NodeResult {
```
Comment (nkunal) [inline pkg/segment/segexecution.go:473]: same comment as above, rename this func based on what name you give for the var above

**Review by nkunal — COMMENTED** [review]
looking good, a few minor nits

File: pkg/segment/query/querystatus.go (hunk 2 above)
Comment (vasukalariya) [reply pkg/segment/query/querystatus.go]: addressed in the latest commit.

File: pkg/segment/query/querystatus.go (hunk 3 above)
Comment (vasukalariya) [reply pkg/segment/query/querystatus.go]: addressed in the latest commit.

File: pkg/segment/query/querystatus.go (hunk 4 above)
Comment (vasukalariya) [reply pkg/segment/query/querystatus.go]: addressed in the latest commit.

File: pkg/segment/query/querystatus.go (hunk 5 above)
Comment (vasukalariya) [reply pkg/segment/query/querystatus.go]: addressed in the latest commit.

File: pkg/segment/segexecution.go (hunk 7 above)
Comment (vasukalariya) [reply pkg/segment/segexecution.go]: addressed in the latest commit.

File: pkg/ast/pipesearch/wsSearchHandler.go [hunk 8]
```diff
@@ -366,6 +377,28 @@ func processMaxScrollComplete(conn *websocket.Conn, qid uint64) {
...
+		wsResponse.GroupByCols = getResp.GroupByCols
+		wsResponse.BucketCount = getResp.BucketCount
+		wsResponse.ColumnsOrder = getResp.ColumnsOrder
+		wsResponse.Qtype = qType.String()
+	case structs.RRCCmd:
+		wsResponse.Hits = getResp.Hits
+		wsResponse.AllPossibleColumns = getResp.AllPossibleColumns
+		wsResponse.Qtype = qType.String()
+		wsResponse.ColumnsOrder = getResp.ColumnsOrder
+	}
```
Comment (AndrewHess) [inline pkg/ast/pipesearch/wsSearchHandler.go:398]: You should have a `default` (and maybe a `case structs.GroupbyCmd`) here, as Lindy mentioned

File: pkg/segment/segexecution.go [hunk 9]
```diff
@@ -459,11 +461,54 @@ func ExecuteAsyncQuery(root *structs.ASTNode, aggs *structs.QueryAggregators, qi
...
+		if config.IsNewQueryPipelineEnabled() {
+			_ = executePipeRespQueryInternal(root, aggs, qid, qc)
+		} else {
+			_ = executeQueryInternal(root, aggs, qid, qc, rQuery)
+		}
 	}()
 	return rQuery.StateChan, nil
 }
 
+func executePipeRespQueryInternal(root *structs.ASTNode, aggs *structs.QueryAggregators, qid uint64, qc *structs.QueryContext) *structs.NodeResult {
```
Comment (AndrewHess) [inline pkg/segment/segexecution.go:473]: I think you copied a lot of this function from the non-websocket searchHander.go. Can you delete it from searchHandler.go now, or is it still different enough that we need duplicate code?

Comment (vasukalariya) [reply pkg/segment/segexecution.go]: addressed in the latest commit.

File: pkg/ast/pipesearch/wsSearchHandler.go (hunk 8 above)
Comment (vasukalariya) [reply pkg/ast/pipesearch/wsSearchHandler.go]: addressed in the latest commit.

**Review by nkunal — COMMENTED** [review]
high level glance looks fine. 
Once @AndrewHess approves, it can be merged

**Review by AndrewHess — APPROVED** [review]

**Pull request merged**